### PR TITLE
feat: supports Gateway resource defined in another namespace

### LINKED ISSUE #1099: Error when running ike develop when the VirtualService is connected to gateway in different namespace
##### Issue Overview

I created a sample application to try istio-workspace and the required k8s and istio component is prepared
- Deployment
- Service
- VirtualService
- DestinationRule

The problem is when we try to run `ike develop` command, it shows this error

<details>
<summary>Error log from controller manager</summary>

```
{
  "error": "failed locating the Gateways that are connected to VirtualServices wallet: failed finding gateway in namespace: Gateway.networking.istio.io \"my_ingress_gateway_from_different_namespace\" not found; failed finding gateway in namespace: Gateway.networking.istio.io \"my_ingress_gateway_from_different_namespace\" not found",
  "errorVerbose": "the following errors occurred:\n -  Gateway.networking.istio.io \"my_ingress_gateway_from_different_namespace\" not found\n    failed finding gateway in namespace\n    github.com/maistra/istio-workspace/pkg/istio.getGateway\n    \t/home/circleci/.go_workspace/src/github.com/maistra/istio-workspace/pkg/istio/gateway.go:196\n    github.com/maistra/istio-workspace/pkg/istio.VirtualServiceGatewayLocator\n    \t/home/circleci/.go_workspace/src/github.com/maistra/istio-workspace/pkg/istio/virtualservicegateway.go:55\n    github.com/maistra/istio-workspace/pkg/model.NewSync.func1\n    \t/home/circleci/.go_workspace/src/github.com/maistra/istio-workspace/pkg/model/sync.go:43\n    github.com/maistra/istio-workspace/controllers/session.(*ReconcileSession).Reconcile\n    \t/home/circleci/.go_workspace/src/github.com/maistra/istio-workspace/controllers/session/session_controller.go:250\n    sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\n    \t/home/circleci/.go_workspace/src/github.com/maistra/istio-workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:121\n    sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n    \t/home/circleci/.go_workspace/src/github.com/maistra/istio-workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:320\n    sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n    \t/home/circleci/.go_workspace/src/github.com/maistra/istio-workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:273\n    sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n    \t/home/circleci/.go_workspace/src/github.com/maistra/istio-workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:234\n    runtime.goexit\n    \t/snap/go/current/src/runtime/asm_amd64.s:1581\n -  Gateway.networking.istio.io \"my_ingress_gateway_from_different_namespace\" not found\n    failed finding gateway in namespace\n    github.com/maistra/istio-workspace/pkg/istio.getGateway\n    \t/home/circleci/.go_workspace/src/github.com/maistra/istio-workspace/pkg/istio/gateway.go:196\n    github.com/maistra/istio-workspace/pkg/istio.VirtualServiceGatewayLocator\n    \t/home/circleci/.go_workspace/src/github.com/maistra/istio-workspace/pkg/istio/virtualservicegateway.go:55\n    github.com/maistra/istio-workspace/pkg/model.NewSync.func1\n    \t/home/circleci/.go_workspace/src/github.com/maistra/istio-workspace/pkg/model/sync.go:43\n    github.com/maistra/istio-workspace/controllers/session.(*ReconcileSession).Reconcile\n    \t/home/circleci/.go_workspace/src/github.com/maistra/istio-workspace/controllers/session/session_controller.go:250\n    sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\n    \t/home/circleci/.go_workspace/src/github.com/maistra/istio-workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:121\n    sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n    \t/home/circleci/.go_workspace/src/github.com/maistra/istio-workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:320\n    sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n    \t/home/circleci/.go_workspace/src/github.com/maistra/istio-workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:273\n    sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n    \t/home/circleci/.go_workspace/src/github.com/maistra/istio-workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:234\n    runtime.goexit\n    \t/snap/go/current/src/runtime/asm_amd64.s:1581\nfailed locating the Gateways that are connected to VirtualServices wallet\ngithub.com/maistra/istio-workspace/pkg/istio.VirtualServiceGatewayLocator\n\t/home/circleci/.go_workspace/src/github.com/maistra/istio-workspace/pkg/istio/virtualservicegateway.go:94\ngithub.com/maistra/istio-workspace/pkg/model.NewSync.func1\n\t/home/circleci/.go_workspace/src/github.com/maistra/istio-workspace/pkg/model/sync.go:43\ngithub.com/maistra/istio-workspace/controllers/session.(*ReconcileSession).Reconcile\n\t/home/circleci/.go_workspace/src/github.com/maistra/istio-workspace/controllers/session/session_controller.go:250\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\n\t/home/circleci/.go_workspace/src/github.com/maistra/istio-workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:121\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/home/circleci/.go_workspace/src/github.com/maistra/istio-workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:320\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/home/circleci/.go_workspace/src/github.com/maistra/istio-workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:273\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/home/circleci/.go_workspace/src/github.com/maistra/istio-workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:234\nruntime.goexit\n\t/snap/go/current/src/runtime/asm_amd64.s:1581",
  "stacktrace": "github.com/maistra/istio-workspace/controllers/session.(*ReconcileSession).Reconcile\n\t/home/circleci/.go_workspace/src/github.com/maistra/istio-workspace/controllers/session/session_controller.go:250\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\n\t/home/circleci/.go_workspace/src/github.com/maistra/istio-workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:121\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/home/circleci/.go_workspace/src/github.com/maistra/istio-workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:320\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/home/circleci/.go_workspace/src/github.com/maistra/istio-workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:273\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/home/circleci/.go_workspace/src/github.com/maistra/istio-workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:234"
}
```
</details>

##### Expected Behaviour

It should have options to provide the namespace for gateways

##### Current Behaviour

It give error as shown above

##### Steps To Reproduce

1. Create the required component (Deployment, Service, VirtualService, DestinationRule) in some namespace
2. Create gateways in different namespace from step 1
4. Assign the gateways to the VirtualService from step 1

### PR DESCRIPTION
**Pull request by bartoszmajsak**
This change enables routing through Gateway defined in another namespace. It requires cluster-wide or multiple ns operator installations (although the latter seems to be discouraged by operator-sdk due to potential tenancy issues).

In addition, it comes with an e2e test demonstrating such a use case.

#### Important note

What this PR doesn't verify yet is `mesh` gateway special case.

Fixes #1099

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -472,19 +472,24 @@ deploy-test-%:
...
 	# Do not remove line breaks as they're intentionally set for docs toolchain to always get right snippet
 	# tag::privileged[]
 	oc adm policy add-scc-to-user privileged -z default -n $(TEST_NAMESPACE) || true
 	# end::privileged[]
+	oc adm policy add-scc-to-user privileged -z default -n $(TEST_GW_NAMESPACE) || true
 
 	go run ./test/cmd/test-scenario/ $(scenario) | $(k8s) apply -f -
 
+
+# TODO FIX undeploying gw if in another ns
```
Comment (bartoszmajsak) [inline Makefile:492]: Coming in the morning.

File: api/maistra/v1alpha1/session_types.go [hunk 2]
```diff
@@ -155,14 +155,24 @@ type Condition struct {
...
-	Name string `json:"name,omitempty"`
-	Ref  string `json:"ref,omitempty"`
-	Kind string `json:"kind,omitempty"`
+	Name      string `json:"name,omitempty"`
+	Namespace string `json:"namespace,omitempty"`
+	Ref       string `json:"ref,omitempty"`
+	Kind      string `json:"kind,omitempty"`
+}
+
+func (s *Source) GetNamespaceName() string {
```
Comment (bartoszmajsak) [inline api/maistra/v1alpha1/session_types.go:164]: @aslakknutsen do we want to keep it?

File: controllers/session/conditions.go [hunk 3]
```diff
@@ -10,16 +10,17 @@ import (
 )
 
 func createConditionForLocatedRef(ref model.Ref, located model.LocatorStatus) istiov1alpha1.Condition {
-	message := located.Kind + "/" + located.Name + " status " + ref.KindName.String() + ": "
+	message := located.GetNamespaceName() + "[" + located.Kind + "] status " + ref.KindName.String() + ": "
```
Comment (bartoszmajsak) [inline controllers/session/conditions.go:13]: I thought this way it's easier to understand when going through `Condition`s in the `Session` object.

File: controllers/session/session_controller.go [hunk 4]
```diff
@@ -103,7 +103,9 @@ func add(mgr manager.Manager, r *ReconcileSession) error {
 	}
 
 	// Watch for changes to primary resource Session
-	err = c.Watch(&source.Kind{Type: &istiov1alpha1.Session{}}, &handler.InstrumentedEnqueueRequestForObject{}, predicate.GenerationChangedPredicate{})
+	err = c.Watch(&source.Kind{Type: &istiov1alpha1.Session{}},
```
Comment (bartoszmajsak) [inline controllers/session/session_controller.go:106]: Small formatting

File: e2e/e2e_suite_test.go [hunk 5]
```diff
@@ -55,7 +55,7 @@ var _ = SynchronizedBeforeSuite(func() []byte {
 
 				<-testshell.Execute(`oc patch configs.imageregistry.operator.openshift.io/cluster --patch '{"spec":{"defaultRoute":true}}' --type=merge`).Done()
 
-				<-testshell.Execute(NewProjectCmd(GetRepositoryName())).Done()
+				<-testshell.Execute(CreateNamespaceCmd(GetRepositoryName())).Done()
```
Comment (bartoszmajsak) [inline e2e/e2e_suite_test.go:58]: I think it better describes what's happening under the hood. `project` is an OpenShift concept.

File: e2e/gw_in_another_ns_test.go [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+		Context("Gateway in another namespace", func() {
+
+			var restoreEnvVars func()
+
+			BeforeEach(func() {
+				scenario = "scenario-1" //nolint:goconst //reason no need for constant (yet)
+
+				namespace = generateNamespaceName()
+				gwNamespace = "gw-" + namespace
```
Comment (bartoszmajsak) [inline e2e/gw_in_another_ns_test.go:37]: Creates two ns, where the one for `Gateway` is prefixed with `gw-`.

File: e2e/gw_in_another_ns_test.go [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+			})
+
+			AfterEach(func() {
+				if CurrentSpecReport().Failed() {
+					PrintFailureDetails(namespace, tmpDir)
+				} else {
+					CleanupNamespace(namespace, false)
+					tmpFs.Cleanup()
+				}
+				<-testshell.Execute(DeleteNamespaceCmd(gwNamespace)).Done()
```
Comment (bartoszmajsak) [inline e2e/gw_in_another_ns_test.go:59]:
```suggestion
				<-testshell.Execute(DeleteNamespaceCmd(namespace)).Done()
				<-testshell.Execute(DeleteNamespaceCmd(gwNamespace)).Done()
```

File: e2e/infra/cluster_setup.go [hunk 8]
```diff
@@ -86,8 +86,8 @@ func UsePrebuiltImages() bool {
 	return os.Getenv("PRE_BUILT_IMAGES") != ""
 }
 
-// PrepareEnv sets up a environmental specific things.
-func PrepareEnv(namespace string) {
+// PrepareEnvForOpenshift sets up an environmental specific things.
+func PrepareEnvForOpenshift(namespace string) {
```
Comment (bartoszmajsak) [inline e2e/infra/cluster_setup.go:90]: I thought this is a better name to explain why this func exists and is called.

File: e2e/infra/istio_operators_setup.go [hunk 9]
```diff
@@ -38,6 +39,20 @@ func InstallLocalOperator(namespace string) {
 	)
 }
 
+func InstallMultiNamespaceOperator(namespace string, watchNs ...string) {
```
Comment (bartoszmajsak) [inline e2e/infra/istio_operators_setup.go:42]: We have to watch multiple ns for "isolated `Gateway`" use case.

File: e2e/infra/test_service.go [hunk 10]
```diff
@@ -42,20 +42,23 @@ func BuildTestServicePreparedImage(callerName string) (registry string) {
...
-		<-shell.ExecuteInDir(".", "bash", "-c",
-			`oc -n `+GetIstioNamespace()+` patch --type='json' smmr default -p '[{"op": "add", "path": "/spec/members/-", "value":"`+namespace+`"}]'`).Done()
-		gomega.Eventually(func() string {
-			return GetProjectLabels(namespace)
-		}, 1*time.Minute).Should(gomega.ContainSubstring("maistra.io/member-of"))
-	} else {
-		shell.WaitForSuccess(
-			shell.ExecuteInDir(".", "bash", "-c", "kubectl label namespace "+namespace+" istio-injection=enabled --overwrite=true"),
-		)
+	for _, ns := range namespace {
```
Comment (bartoszmajsak) [inline e2e/infra/test_service.go:48]: Now we need to iterate over all ns involved and make them "istio-fied".

File: pkg/istio/gateway.go [hunk 11]
```diff
@@ -134,17 +134,17 @@ func mutateGateway(ctx model.SessionContext, source istionetwork.Gateway) (mutat
 		hosts := server.Hosts
 		for _, host := range hosts {
 			newHost := ctx.Name + "." + host
-			if !existInList(existingHosts, host) && !existInList(existingHosts, newHost) {
+			if !isInSlice(existingHosts, host) && !isInSlice(existingHosts, newHost) {
```
Comment (bartoszmajsak) [inline pkg/istio/gateway.go:137]: I thought it was a better name.

File: pkg/istio/virtualservice.go [hunk 12]
```diff
@@ -487,13 +488,14 @@ func getHostsFromGateway(ctx model.SessionContext, store model.LocatorStatusStor
 	gwByName := func(store model.LocatorStatusStore, gatewayName string) []model.LocatorStatus {
 		var f []model.LocatorStatus
 		for _, g := range store(GatewayKind) {
-			if g.Name == gatewayName {
+			if g.GetNamespaceName() == gatewayName || g.Name == gatewayName {
```
Comment (bartoszmajsak) [inline pkg/istio/virtualservice.go:491]: That's one of the key changes to make this thing work.

File: pkg/istio/virtualservicegateway.go [hunk 13]
```diff
@@ -51,7 +51,18 @@ func VirtualServiceGatewayLocator(ctx model.SessionContext, ref model.Ref, store
 			vs := vss.Items[i]
 			if gateways, connected := connectedToGateway(vs); connected {
 				for _, gwName := range gateways {
-					gw, err := getGateway(ctx, ctx.Namespace, gwName)
+					gwName := gwName // pin
```
Comment (bartoszmajsak) [inline pkg/istio/virtualservicegateway.go:54]: That's one of the key changes to make this thing work.

Comment (bartoszmajsak) [reply pkg/istio/virtualservicegateway.go:54]: I mean lines below :)

File: e2e/gw_in_another_ns_test.go (hunk 7 above)
Comment (bartoszmajsak) [reply e2e/gw_in_another_ns_test.go:59]: It's not needed - `CleanupNamespace` does that already

**Review by aslakknutsen — COMMENTED** [review]
I see a few places where we Create resources, eg. vs or dr, we use ctx.Namespace for the new Resource namespace when we have access to the Ref model. I don't think it matters much beyond; should we be more consistent and rely on the Ref model Namespace instead of ctx where possible?

Basically let the Locators mainly deal with the correct namespace, the 'Actions' just rely on the info provided in the Ref model.

File: e2e/infra/cluster_setup.go (hunk 8 above)
Comment (bartoszmajsak) [reply e2e/infra/cluster_setup.go:90]: It's the only thing going on there now.

**Review by aslakknutsen — APPROVED** [review]

**Pull request merged**